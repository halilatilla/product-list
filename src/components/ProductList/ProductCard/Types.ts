import { IProduct } from '@src/types'

export interface IProps {
  className?: string
  product: IProduct
}
