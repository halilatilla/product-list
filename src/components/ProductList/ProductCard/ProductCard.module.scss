.productCard {
  border-radius: 8px;
  height: 464px;
  border: 1px solid transparent;
  display: grid;
  grid-template-rows: max-content 1fr;

  &:hover {
    border-color: var(--c-border-dark);
    padding-bottom: 24px;

    .imageWrapper {
      border-color: transparent;
    }

    .info {
      display: none;
    }

    .addToCart {
      display: block;
    }
  }
}

.imageWrapper {
  border: 1px solid var(--c-border-light);
  width: 100%;
  border-radius: 8px;
}

.image {
  background-size: cover;
  background-position: center;
}

.detail {
  padding: 8px 8px 0 8px;
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  font-size: var(--font-size-xs);
  text-transform: capitalize;
}

.subTitle {
  text-transform: capitalize;
  font-weight: bold;
}

.priceWrapper {
  margin-top: 8px;
}

.price {
  font-weight: var(--font-weight-bold);
  font-size: var(--font-size-sm);
  color: var(--c-black);
}

.discountPrice {
  font-size: 13px;
  color: var(--c-text-disabled);
  text-decoration: line-through;
  margin-right: 4px;
}

.discountPercent {
  font-weight: var(--font-weight-bold);
  color: var(--c-primary);
}

.addToCart {
  display: none;
}
