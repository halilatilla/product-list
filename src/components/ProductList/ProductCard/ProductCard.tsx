import { FC, useState, useEffect } from 'react'
import classnames from 'classnames'
import Image from 'next/image'

import { getPriceDiscount } from '@src/lib'
import { useAppDispatch, addToCart, useAppSelector } from '@src/store'
import { Button } from '@src/components'
import IProductCard from './ProductCard.types'
import styles from './ProductCard.module.scss'

const ProductCard: FC<IProductCard> = ({ className, product, ...rest }) => {
  const { productId, imgUrl, title, brand, color, discount, price } = product

  const [itemIsAlreadyAdded, setItemIsAlreadyAdded] = useState(false)

  const store = useAppSelector((state) => state.cart)

  const dispatch = useAppDispatch()

  const handleAddToCart = () => {
    dispatch(addToCart({ productId, imgUrl, title }))
  }

  const getItemIsAlreadyAdded = () => {
    const result = store.some((item) => item.productId === productId)
    if (result) {
      setItemIsAlreadyAdded(true)
    } else {
      setItemIsAlreadyAdded(false)
    }
  }

  useEffect(() => {
    getItemIsAlreadyAdded()
  }, [store])

  return (
    <li className={classnames(styles.productCard, className)} {...rest}>
      <div className={classnames(styles.imageWrapper, 'flex-center')}>
        <Image src={imgUrl} alt={title} className={styles.image} width={224} height={332} />
      </div>
      <div className={styles.detail}>
        <p className={styles.title}> {title} </p>
        <div className={styles.info}>
          <p className={styles.brand}>
            <span className={styles.subTitle}>marka:</span> <span>{brand}</span>
          </p>
          <p className={styles.color}>
            <span className={styles.subTitle}>renk:</span>
            <span>{color}</span>
          </p>
          <div className={styles.priceWrapper}>
            <p className={styles.price}> {getPriceDiscount(parseFloat(price), discount)} TL </p>
            <div>
              <span className={styles.discountPrice}>{price}</span>
              <span className={styles.discountPercent}> {discount}% </span>
            </div>
          </div>
        </div>
        {itemIsAlreadyAdded ? (
          <Button
            onClick={handleAddToCart}
            className={styles.addToCart}
            label="bu ürünü sepete ekleyemezsiniz"
            appearance="addToCart"
            disabled
          />
        ) : (
          <Button onClick={handleAddToCart} className={styles.addToCart} label="sepete ekle" appearance="addToCart" />
        )}
      </div>
    </li>
  )
}

export default ProductCard
