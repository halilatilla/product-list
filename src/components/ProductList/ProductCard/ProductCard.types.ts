import { IProduct } from '@src/types'

export default interface IProductCard {
  className?: string
  product: IProduct
}
