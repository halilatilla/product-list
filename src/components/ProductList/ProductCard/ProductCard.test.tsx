import { render } from '@src/testUtils'

import ProductCard from './ProductCard'

jest.mock('next/image', () => ({
  // eslint-disable-next-line @typescript-eslint/naming-convention
  __esModule: true,
  default: (props: any) => {
    // eslint-disable-next-line jsx-a11y/alt-text
    return <img width={100} height={100} {...props} />
  },
}))

describe('ProductCard', () => {
  const PROPS = {
    product: {
      productId: '',
      imgUrl: '/',
      title: '',
      brand: '',
      color: '',
      discount: 0,
      price: '',
      createdDate: '',
    },
  }

  it('renders correctly', () => {
    const { container } = render(<ProductCard {...PROPS} />)
    expect(container).toMatchSnapshot()
  })
})
