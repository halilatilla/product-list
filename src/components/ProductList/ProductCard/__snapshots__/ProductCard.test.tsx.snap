// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductCard renders correctly 1`] = `
<div>
  <li
    class="productCard"
  >
    <div
      class="imageWrapper flex-center"
    >
      <span
        style="box-sizing: border-box; display: inline-block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: relative; max-width: 100%;"
      >
        <span
          style="box-sizing: border-box; display: block; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; max-width: 100%;"
        >
          <img
            alt=""
            aria-hidden="true"
            src="data:image/svg+xml;base64,PHN2ZyB3aWR0aD0iMjI0IiBoZWlnaHQ9IjMzMiIgeG1sbnM9Imh0dHA6Ly93d3cudzMub3JnLzIwMDAvc3ZnIiB2ZXJzaW9uPSIxLjEiLz4="
            style="display: block; max-width: 100%; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px;"
          />
        </span>
        <img
          alt=""
          class="image"
          data-nimg="intrinsic"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="detail"
    >
      <p
        class="title"
      >
         
        
         
      </p>
      <div
        class="info"
      >
        <p
          class="brand"
        >
          <span
            class="subTitle"
          >
            marka:
          </span>
           
          <span />
        </p>
        <p
          class="color"
        >
          <span
            class="subTitle"
          >
            renk:
          </span>
          <span />
        </p>
        <div
          class="priceWrapper"
        >
          <p
            class="price"
          >
             
            NaN
             TL 
          </p>
          <div>
            <span
              class="discountPrice"
            />
            <span
              class="discountPercent"
            >
               
              0
              % 
            </span>
          </div>
        </div>
      </div>
      <button
        class="button flex-center addToCart addToCart"
        type="button"
      >
        sepete ekle
         
      </button>
    </div>
  </li>
</div>
`;
